pkg/server/server.go

```diff
@@ -46,46 +51,109 @@ var tracer = otel.Tracer("openfga/pkg/server")
-	Datastore    storage.OpenFGADatastore
-	Logger       logger.Logger
-	Transport    gateway.Transport
-	TokenEncoder encoder.Encoder
+type OpenFGASeviceV1Option func(s *Server)
+
+func WithDatastore(ds storage.OpenFGADatastore) OpenFGASeviceV1Option {
+	return func(s *Server) {
+		s.datastore = ds
+	}
+}
+
+func WithLogger(l logger.Logger) OpenFGASeviceV1Option {
+	return func(s *Server) {
+		s.logger = l
+	}
+}
+
+func WithTokenEncoder(encoder encoder.Encoder) OpenFGASeviceV1Option {
+	return func(s *Server) {
+		s.encoder = encoder
+	}
+}
+
+func WithTransport(t gateway.Transport) OpenFGASeviceV1Option {
+	return func(s *Server) {
+		s.transport = t
+	}
+}
+
+func WithResolveNodeLimit(limit uint32) OpenFGASeviceV1Option {
+	return func(s *Server) {
+		s.resolveNodeLimit = limit
+	}
+}
+
+func WithChangelogHorizonOffset(offset int) OpenFGASeviceV1Option {
+	return func(s *Server) {
+		s.changelogHorizonOffset = offset
+	}
+}
+
+func WithListObjectsDeadline(deadline time.Duration) OpenFGASeviceV1Option {
+	return func(s *Server) {
+		s.listObjectsDeadline = deadline
+	}
+}
+
+func WithListObjectsMaxResults(limit uint32) OpenFGASeviceV1Option {
+	return func(s *Server) {
+		s.listObjectsMaxResults = limit
+	}
+}
+
+func WithExperimentals(experimentals ...ExperimentalFeatureFlag) OpenFGASeviceV1Option {
+	return func(s *Server) {
+		s.experimentals = experimentals
+	}
 }
 
-type Config struct {
-	ResolveNodeLimit       uint32
-	ChangelogHorizonOffset int
-	ListObjectsDeadline    time.Duration
-	ListObjectsMaxResults  uint32
-	Experimentals          []ExperimentalFeatureFlag
+func MustNewServerWithOpts(opts ...OpenFGASeviceV1Option) *Server {
+	s, err := NewServerWithOpts(opts...)
+	if err != nil {
+		panic(fmt.Sprintf("failed to construct the OpenFGA server: %v", err))
+	}
+
+	return s
 }
 
-// New creates a new Server which uses the supplied backends
-// for managing data.
-func New(dependencies *Dependencies, config *Config) *Server {
+func NewServerWithOpts(opts ...OpenFGASeviceV1Option) (*Server, error) {
 
-	optimizeListObjects := false
-	if slices.Contains(config.Experimentals, optimizedListObjects) {
-		optimizeListObjects = true
+	s := &Server{
+		logger:                logger.NewNoopLogger(),
+		encoder:               encoder.NewBase64Encoder(),
+		transport:             gateway.NewNoopTransport(),
+		resolveNodeLimit:      defaultResolveNodeLimit,
+		listObjectsDeadline:   defaultListObjectsDeadline,
+		listObjectsMaxResults: defaultListObjectsMaxResults,
+		optimizeListObjects:   false,
 	}
 
-	return &Server{
-		logger:              dependencies.Logger,
-		datastore:           dependencies.Datastore,
-		encoder:             dependencies.TokenEncoder,
-		transport:           dependencies.Transport,
-		config:              config,
-		optimizeListObjects: optimizeListObjects,
+	for _, opt := range opts {
+		opt(s)
 	}
+
+	if s.datastore == nil {
+		return nil, fmt.Errorf("a datastore option must be provided")
+	}
```

Comment:
Please add a test

```go
func TestPanicIfNoDatastore(t *testing.T) {
	require.PanicsWithError(t, "failed to construct the OpenFGA server: a datastore option must be provided", func() {
		_ = MustNewServerWithOpts()
	})
}
```

Reply:
Added in [b090bdd](https://github.com/openfga/openfga/pull/833/commits/b090bdd9c4233375ab879a8ae5f5d4acb63f3f0f)